compat/mingw.c

```diff
@@ -426,6 +426,67 @@ static wchar_t *normalize_ntpath(wchar_t *wbuf)
 	return wbuf;
 }
 
+static BOOL exists(const wchar_t *wpathname)
+{
+	HANDLE hnd = CreateFileW(wpathname, GENERIC_READ,
+				 FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE, NULL,
+				 OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
+	if (hnd == INVALID_HANDLE_VALUE)
+		return FALSE;
+	CloseHandle(hnd);
+	return TRUE;
+}
+
+static BOOL is_exclusively_writable(const wchar_t *wpathname)
+{
+	HANDLE hnd = CreateFileW(wpathname, GENERIC_WRITE, 0, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
+	if (hnd == INVALID_HANDLE_VALUE)
+		return !is_file_in_use_error(GetLastError());
+	CloseHandle(hnd);
+	return TRUE;
+}
+
+/* rename and unlink a file. This is needed for the ability to recreate a file
+ * with the same name, for example on checkout which unlinks and recreates the file */
+static int rename_and_unlink(const wchar_t *wpathname, int (*unlinker)(const wchar_t *))
+{
+	wchar_t wtemppath[MAX_LONG_PATH];
+	wchar_t *last, *original;
+	*wtemppath = 0;
+	/* leave some space for the .unlink extension */
+	wcsncat(wtemppath, wpathname, sizeof(wtemppath) - 8);
+	original = last = wtemppath + wcslen(wtemppath);
+	/* Start with file.unlink. If it exists, append X up to 12 times */
+	wcscpy(last, L".unlink");
+	last += (sizeof(L".unlink") / sizeof (wchar_t)) - 1;
+	while (exists(wtemppath)) {
+		if ((last - original >= 20) || (last - wtemppath + 1 >= MAX_LONG_PATH)) {
+			last = original;
+			*last = 0;
+			break;
+		}
+		*last++ = L'X';
+		*last = 0;
+	}
+	if (last > original) {
+		int res = 0;
+		if (MoveFileExW(wpathname, wtemppath, 0)) {
+			res = unlinker(wtemppath);
+			if (res)
+				MoveFileExW(wtemppath, wpathname, 0);
+		}
+		return res;
+	}
+	return unlinker(wpathname);
+}
+
+static int do_unlink(const wchar_t *wpathname, int (*unlinker)(const wchar_t *))
+{
+	if (is_exclusively_writable(wpathname))
```

Comment:
If the result of the `is_exclusively_writable` check can change between this if condition and our call to `unlink`, then this is a race condition. I think reliable behavior is better. How about this: Keep using `unlink` by default, but add a configuration option for Qt Creator users to always use `rename_and_unlink` (we can mention the potential performance impact in the docs).

Reply:
Ok, so I can try to use FILE_SHARE_DELETE here, and leave the handle open until after the actual unlink.

Reply:
Done